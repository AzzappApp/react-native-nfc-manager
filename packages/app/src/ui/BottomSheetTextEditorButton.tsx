import { colors } from '#theme';
import { createStyleSheet, useStyleSheet } from '#helpers/createStyles';

import IconButton from './IconButton';
import type { TextAndSelection } from './BottomSheetTextEditorTypes';
import type { Icons } from './Icon';
import type { RichTextASTTags } from '@azzapp/shared/richText/richTextTypes';

const defaultIconSize = 12;

type BottomSheetTextEditorButtonProps = {
  onPress: (arg: RichTextASTTags) => void;
  textAndSelection: TextAndSelection;
  tag: RichTextASTTags;
  icon: Icons;
  isFocused: boolean;
  iconSize?: number;
  enabledButtons?: RichTextASTTags[];
};

export const BottomSheetTextEditorButton = ({
  onPress,
  textAndSelection,
  tag,
  icon,
  isFocused,
  enabledButtons,
  iconSize = defaultIconSize,
}: BottomSheetTextEditorButtonProps) => {
  const styles = useStyleSheet(styleSheet);

  if (enabledButtons && !enabledButtons.includes(tag)) {
    return undefined;
  }
  const isDisableInner =
    !isFocused ||
    textAndSelection.selection?.start === textAndSelection.selection?.end;

  return (
    <IconButton
      icon={icon}
      iconSize={iconSize}
      variant="border"
      onPress={() => onPress(tag)}
      iconStyle={isDisableInner ? styles.buttonDisable : undefined}
      style={[
        styles.button,
        textAndSelection.selectedTag.find(t => t === tag) === undefined
          ? undefined
          : styles.buttonSelected,
      ]}
      disabled={isDisableInner}
      disabledOpacity={1}
    />
  );
};

const styleSheet = createStyleSheet(appearance => ({
  button: {
    width: 54,
    borderRadius: 33,
    borderColor: appearance === 'dark' ? colors.grey900 : colors.grey50,
    borderWidth: 1,
    height: '100%',
  },
  buttonDisable: {
    tintColor: appearance === 'dark' ? colors.grey800 : colors.grey200,
  },
  buttonSelected: {
    backgroundColor: appearance === 'dark' ? colors.grey900 : colors.grey50,
  },
}));

export default BottomSheetTextEditorButton;
