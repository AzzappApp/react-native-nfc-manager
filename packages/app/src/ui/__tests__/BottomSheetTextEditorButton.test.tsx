import { render, screen } from '@testing-library/react-native';
import BottomSheetTextEditorButton from '#ui/BottomSheetTextEditorButton';

describe('BottomSheetTextEditorButton component', () => {
  test('selected state', () => {
    render(
      <BottomSheetTextEditorButton
        onPress={() => {}}
        textAndSelection={{
          ast: {
            type: 'root',
            start: 0,
            end: 1,
            children: undefined,
            value: undefined,
          },
          selection: undefined,
          selectedTag: ['b'],
        }}
        tag="b"
        icon="link"
        isFocused
      />,
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });

  test('unselected state', () => {
    render(
      <BottomSheetTextEditorButton
        onPress={() => {}}
        textAndSelection={{
          ast: {
            type: 'root',
            start: 0,
            end: 1,
            children: undefined,
            value: undefined,
          },
          selection: undefined,
          selectedTag: [],
        }}
        tag="b"
        icon="link"
        isFocused
      />,
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });

  test('disabled state', () => {
    render(
      <BottomSheetTextEditorButton
        onPress={() => {}}
        textAndSelection={{
          ast: {
            type: 'root',
            start: 0,
            end: 1,
            children: undefined,
            value: undefined,
          },
          selection: undefined,
          selectedTag: [],
        }}
        tag="b"
        icon="link"
        isFocused={false}
      />,
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
