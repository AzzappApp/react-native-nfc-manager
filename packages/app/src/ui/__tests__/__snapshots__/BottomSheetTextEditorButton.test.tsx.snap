// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomSheetTextEditorButton component disabled state 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      [
        {
          "aspectRatio": 1,
          "height": 50,
          "minWidth": 50,
        },
        [
          {
            "borderColor": "#F5F5F6",
            "borderRadius": 33,
            "borderWidth": 1,
            "height": "100%",
            "width": 54,
          },
          undefined,
        ],
      ],
      {
        "opacity": 1,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderColor": "#000000",
          "borderWidth": 1,
          "flex": 1,
          "justifyContent": "center",
          "overflow": "hidden",
          "width": "100%",
        },
        {
          "borderRadius": 25,
        },
      ]
    }
  >
    <Image
      source={
        {
          "testUri": "../../../packages/app/src/ui/Icon/assets/link.png",
        }
      }
      style={
        [
          {
            "height": 12,
            "resizeMode": "contain",
            "width": 12,
          },
          {
            "tintColor": "#0E1216",
          },
          {
            "tintColor": "#C8C7CA",
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`BottomSheetTextEditorButton component selected state 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      [
        {
          "aspectRatio": 1,
          "height": 50,
          "minWidth": 50,
        },
        [
          {
            "borderColor": "#F5F5F6",
            "borderRadius": 33,
            "borderWidth": 1,
            "height": "100%",
            "width": 54,
          },
          {
            "backgroundColor": "#F5F5F6",
          },
        ],
      ],
      {
        "opacity": 1,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderColor": "#000000",
          "borderWidth": 1,
          "flex": 1,
          "justifyContent": "center",
          "overflow": "hidden",
          "width": "100%",
        },
        {
          "borderRadius": 25,
        },
      ]
    }
  >
    <Image
      source={
        {
          "testUri": "../../../packages/app/src/ui/Icon/assets/link.png",
        }
      }
      style={
        [
          {
            "height": 12,
            "resizeMode": "contain",
            "width": 12,
          },
          {
            "tintColor": "#0E1216",
          },
          {
            "tintColor": "#C8C7CA",
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`BottomSheetTextEditorButton component unselected state 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      [
        {
          "aspectRatio": 1,
          "height": 50,
          "minWidth": 50,
        },
        [
          {
            "borderColor": "#F5F5F6",
            "borderRadius": 33,
            "borderWidth": 1,
            "height": "100%",
            "width": 54,
          },
          undefined,
        ],
      ],
      {
        "opacity": 1,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderColor": "#000000",
          "borderWidth": 1,
          "flex": 1,
          "justifyContent": "center",
          "overflow": "hidden",
          "width": "100%",
        },
        {
          "borderRadius": 25,
        },
      ]
    }
  >
    <Image
      source={
        {
          "testUri": "../../../packages/app/src/ui/Icon/assets/link.png",
        }
      }
      style={
        [
          {
            "height": 12,
            "resizeMode": "contain",
            "width": 12,
          },
          {
            "tintColor": "#0E1216",
          },
          {
            "tintColor": "#C8C7CA",
          },
        ]
      }
    />
  </View>
</View>
`;
